docs/kb/pcr-measurements.md

```diff
@@ -0,0 +1,157 @@
+
+## Introduction
+
+As part of measured boot process firmware hashes (measures) various
+pieces of code or data and updates PCRs of a TPM device.  This allows a user
+(typically, automatically) to attest integrity of the system by verifying PCR
+values after a boot or tying decryption of data to expected values.
+
+Usually the firmware is not the only entity which updates PCRs and even if it
+is, the number of updates can be large, so it's important to know which values
+do get measured to which PCRs and under what conditions (e.g., some
+measurements are done in response to user actions).
+
+TPM event log maintained throughout the boot process is meant to track this kind
+of information, but due to limitations of its format the log is rarely enough
+to understand all of the measurements.  This document is meant to describe when
+Dasharo updates PCRs so users would know which PCRs to use and when to expect
+their values to be changed.
+
+Firmware components are described separately because not all components might
+be present in a given firmware variant.
+
+## The truth about PCR usage
+
+[TPM 1.2 Specification][TPM12] (section 3.3.3) and
+[TPM 2.0 Specification][TPM20] (section 3.3.4) provide information on which
+PCRs are expected to be used for what purposes.  The problem is that
+instructions seemingly contradict themselves and even if when they don't,
+discerning which PCR is to be used for a particular code or data is far from
+an easy task.  This results in a disconnect between what's prescribed by those
+specification and how things are in practice.
+
+EDK is better at adhering to TCG specifications probably because of interactions
+between developers on both sides.  The way coreboot updates PCRs originally came
+from Chromebooks and its authors had a different interpretation of PCR
+allocations ([see][coreboot-pcrs]).  When EDK is used as a second-stage
+firmware after coreboot, you get a mix of both.
+
+[TPM12]: https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientImplementation_1-21_1_00.pdf
+[TPM20]: https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf
+[coreboot-pcrs]: https://doc.coreboot.org/security/vboot/measured_boot.html#platform-configuration-registers
+
+## coreboot measurements
+
+### PCR banks
+
+At the time of writing (June 2024) coreboot has no support for multiple PCR
+banks but allows choosing a hash algorithm for TPM 2.0 log format.  This choice
+is fixed at build time.
+
+| TPM version | PCR Banks
+| ----------- | --------
+| 1.2         | SHA-1 only
+| 2.0         | SHA-256 only
+
+This has an implication for TPM 2.0 case when multiple PCR banks are active and
+used by EDK.  This is handled by using dummy hash values (`0100...`) for missing
+digests (SHA-1) which do not correspond to actual PCR updates.  Doing this
+breaks replaying of TPM log unless fake digests are skipped, but it allows to
+comply with TPM event log format.
```

Comment:
This is done exactly to mark that given bank isn't used. On top of that, a dummy value (`EV_SEPARATOR`, which I believe is `0100...` mentioned in this paragraph) should be extended on PCRs from unused banks before execution is passed to user-controlled environment. Without it, an attacker could fake the PCR values by extending them from their defaults of all 0s, by replaying the event log on live PCRs.

If `EV_SEPARATOR` exists in the event log, but isn't extended in actual PCRs, it is a bug. See `8.2.4 Measuring OS Boot Events` in TPM2.0 spec linked earlier in the doc.

Reply:
Check description in Table 14, it has two separate cases for error and success for `EV_SEPARATOR`, `0100...` seems to be the error case, after endianess swap.

Convenience shouldn't take precedence over abiding to specification. It is better to make log replaying fail in this case, as it suggests security issue (non-capped PCRs that can be extended _after_ the chain was already broken, for given bank).

In any case, log by itself isn't a proof, you can't depend on its correctness unless root of trust isn't compromised **and** PCR values are known to be good, and not only matching the log. In other words, if PCRs match the event log, it means that event log is properly created, but it doesn't give any security guarantees. Here, we don't even match the "properly created" criteria.

Reply:
> Check description in Table 14, it has two separate cases for error and success for EV_SEPARATOR, 0100... seems to be the error case, after endianess swap.

That is still about capping of PCRs, it's fixed to 4 bytes and I don't think endianness applies because measured data is a byte array.

> (non-capped PCRs that can be extended after the chain was already broken, for given bank).

It's capped by EDK which also adds fake digests to the log.